File: 2023/day06/clojure/solution.clj
(ns solution
  (:require [clojure.string :as str]))

(defn read-input []
  (slurp "../input.txt"))

(defn parse-numbers [line]
  (->> (str/split line #":")
       second
       str/trim
       (#(str/split % #"\s+"))
       (map parse-long)))

(defn parse-races [input]
  (let [lines (str/split-lines (str/trim input))
        times (parse-numbers (first lines))
        distances (parse-numbers (second lines))]
    (map vector times distances)))

(defn count-ways-to-win [time record]
  "Count the number of ways to beat the record using the quadratic formula.

   If we hold the button for t ms, we travel t * (time - t) mm.
   We need: t * (time - t) > record
   Solving: -t^2 + time*t - record > 0
   Roots: t = (time +/- sqrt(time^2 - 4*record)) / 2"
  (let [discriminant (- (* time time) (* 4 record))]
    (if (<= discriminant 0)
      0
      (let [sqrt-d (Math/sqrt discriminant)
            t-low (/ (- time sqrt-d) 2)
            t-high (/ (+ time sqrt-d) 2)
            ;; We need integer values strictly between the roots
            first-val (inc (long (Math/floor t-low)))
            last-val (dec (long (Math/ceil t-high)))]
        (if (< last-val first-val)
          0
          (inc (- last-val first-val)))))))

(defn part1 [input]
  (let [races (parse-races input)]
    (reduce * (map (fn [[time record]] (count-ways-to-win time record)) races))))

(defn part2 [input]
  (let [races (parse-races input)
        time (parse-long (apply str (map first races)))
        record (parse-long (apply str (map second races)))]
    (count-ways-to-win time record)))

(defn -main []
  (let [input (read-input)]
    (println "Part 1:" (part1 input))
    (println "Part 2:" (part2 input))))

(-main)

File: 2023/day06/rust/solution.rs
use std::fs;
use std::path::Path;

fn count_ways_to_win(time: u64, record: u64) -> u64 {
    // If we hold the button for t ms, we travel t * (time - t) mm.
    // We need: t * (time - t) > record
    // Solving: -t^2 + time*t - record > 0
    // Roots: t = (time +/- sqrt(time^2 - 4*record)) / 2

    let time_f = time as f64;
    let record_f = record as f64;

    let discriminant = time_f * time_f - 4.0 * record_f;
    if discriminant <= 0.0 {
        return 0;
    }

    let sqrt_d = discriminant.sqrt();
    let t_low = (time_f - sqrt_d) / 2.0;
    let t_high = (time_f + sqrt_d) / 2.0;

    // We need integer values strictly between the roots
    let first = t_low.floor() as u64 + 1;
    let last = t_high.ceil() as u64 - 1;

    if last < first {
        return 0;
    }
    last - first + 1
}

fn parse_races(input: &str) -> Vec<(u64, u64)> {
    let lines: Vec<&str> = input.lines().collect();

    let times: Vec<u64> = lines[0]
        .split(':')
        .nth(1)
        .unwrap()
        .split_whitespace()
        .map(|s| s.parse().unwrap())
        .collect();

    let distances: Vec<u64> = lines[1]
        .split(':')
        .nth(1)
        .unwrap()
        .split_whitespace()
        .map(|s| s.parse().unwrap())
        .collect();

    times.into_iter().zip(distances).collect()
}

fn part1(input: &str) -> u64 {
    let races = parse_races(input);
    races
        .iter()
        .map(|&(time, record)| count_ways_to_win(time, record))
        .product()
}

fn part2(input: &str) -> u64 {
    let races = parse_races(input);

    // Concatenate all times and distances into single numbers
    let time_str: String = races.iter().map(|(t, _)| t.to_string()).collect();
    let distance_str: String = races.iter().map(|(_, d)| d.to_string()).collect();

    let time: u64 = time_str.parse().unwrap();
    let record: u64 = distance_str.parse().unwrap();

    count_ways_to_win(time, record)
}

fn main() {
    let input_path = Path::new(file!()).parent().unwrap().join("../input.txt");
    let input = fs::read_to_string(&input_path)
        .expect("Failed to read input file")
        .trim()
        .to_string();

    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}
